### PR TITLE
add tests for torch.testing

### PR DESCRIPTION
**Pull request by pmeier**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #56365 Split `test_testing.py::TestAsserts` for multiple devices
* #55890 enable support for custom error messages in `torch.testing`
* #55788 add support for equal_nan in torch.testing.assert_close
* #55786 Add support for scalars and numpy in torch.testing
* #55385 Add support for checking tensor containers in `torch.testing`
* #55145 make torch.testing error messages more expressive
* **#54784 add tests for torch.testing**

Differential Revision: [D27717422](https://our.internmc.facebook.com/intern/diff/D27717422)

### CODE REVIEW

File: test/test_testing.py [hunk 1]
```diff
@@ -683,5 +685,185 @@ def test_calculate_2_shards_against_optimal_shards(self):
                 self.assertEqual(sorted_tests, sorted_shard_tests)
 
 
+class TestAsserts(TestCase):
+    @staticmethod
+    def _assert_fns(device=None):
+        return [torch.testing.assert_tensors_equal, torch.testing.assert_tensors_close]
```
Comment (pmeier) [inline test/test_testing.py:691]: Not sure why the device parameter is needed here. Is there a better way to add helper functions to a test case that will be used in `instantiate_device_type_tests()`?

Comment (mruberry) [reply test/test_testing.py:691]: You shouldn't need the device parameter here. Here's an example helper method in another device generic test suite:

https://github.com/pytorch/pytorch/blob/f5d6b90c3573a0e5507e4e100569924f6b2ba7f4/test/test_unary_ufuncs.py#L248

**Review by mruberry — APPROVED** [review]
Tests cases look good. I look forward to more thorough tests like these:

https://github.com/pytorch/pytorch/blob/f5d6b90c3573a0e5507e4e100569924f6b2ba7f4/test/test_testing.py#L110

which test for detailed error messages and more complicated comparisons.

Ping me when the helper method issue is sorted out and let's get this merged.

File: test/test_testing.py (hunk 1 above)
Comment (pmeier) [reply test/test_testing.py:691]: This helper probably has the exact same problem as I have. It has the `dtype` parameter without actually using it. If I remove it and try to run the test case, I'm hit with

```
Traceback (most recent call last):
  File "/home/philip/git/pytorch/torch/test/test_unary_ufuncs.py", line 353, in test_reference_numerics_extremal
    self._test_reference_numerics(dtype, op, tensors, equal_nan)
  File "/home/philip/git/pytorch/torch/test/test_unary_ufuncs.py", line 309, in _test_reference_numerics
    _helper_reference_numerics(expected, actual, msg, exact_dtype, equal_nan)
  File "/home/philip/git/pytorch/torch/test/test_unary_ufuncs.py", line 287, in _helper_reference_numerics
    self.assertEqualHelper(actual, expected, msg, dtype=dtype, equal_nan=equal_nan, exact_dtype=exact_dtype)
  File "/home/philip/git/pytorch/torch/test/test_unary_ufuncs.py", line 270, in assertEqualHelper
    **kwargs)
TypeError: assertEqual() got an unexpected keyword argument 'dtype'
```

I suspect the test creation simply adds the `dtype` (or `device`) parameter to every method instead of just the ones prefixed by `_test`. I'll investigate.

**Comment by pmeier** [comment]
> Tests cases look good. I look forward to more thorough tests like these:
> 
> https://github.com/pytorch/pytorch/blob/f5d6b90c3573a0e5507e4e100569924f6b2ba7f4/test/test_testing.py#L110
>
> which test for detailed error messages and more complicated comparisons.

I like the general idea, but I wouldn't be as strict as the old tests. They assert that exactly this message is raised. If we ever change the wording slightly we need to adapt each test. I suggest we check for parts of the message instead, for example is the index of the max diff included.

**Comment by mattip** [comment]
> If we ever change the wording slightly we need to adapt each test.

It makes sense to me to test only the part of the message that indicates the error came from the stanza that raised it. Otherwise (in general, perhaps not applicable in this case) we may find that changing a version of python/numpy/glibc makes tests fail.

File: test/test_testing.py (hunk 1 above)
Comment (pmeier) [reply test/test_testing.py:691]: Scratch the above. The problem is that the method is defined as `@staticmethod` but the instantiation adds it back as normal.

**Comment by pmeier** [comment]
The error messages for mismatched values are directly adapted from the old `_compare_tensors_internal`:

https://github.com/pytorch/pytorch/blob/50cb75edcef41b82fefe957f95b470dfffe53710/torch/testing/__init__.py#L104-L111


A sample error message right now looks like this

```
With rtol=1.3e-06 and atol=1e-05, found 1 different element(s) out of 12 (8.3%). The greatest difference of 4.0 (5.0 vs. 9.0) occurred at index (2, 3)
```

Using the same data with `numpy.testing.assert_equal` gives the following output:

```
Not equal to tolerance rtol=1.3e-06, atol=1e-05

Mismatched elements: 1 / 12 (8.33%)
Max absolute difference: 4.
Max relative difference: 0.44444445
 x: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 5.]], dtype=float32)
 y: array([[5., 5., 5., 5.],
       [5., 5., 5., 5.],
       [5., 5., 5., 9.]], dtype=float32)
```

Pros:

- The info is presented in a list instead of a sentence. IMO this makes it more readable
- The maximum relative difference is reported, which is beneficial in case a comparison fails due to the `rtol`

Cons:

- The values of the inputs are reported (this can be disabled by passing `verbose=False`, but lets face it: most users will use the default setting). In case the inputs are large, the output gets truncated with `...`. Not only is it hard to visually find the mismatching values, they could also live within the truncated part, making the output completely useless.
- Even when visually find the offending values it is hard to parse this back to the index in the inputs.

I propose we use a mix of both:

```
Tensors are not close:

Mismatched elements: 1 / 12 (8.3%)
Max. diff. index: [2, 3]
Max. rel. diff.: 4.44e-1 (rtol=1.3e-06)
Max. abs. diff.: 4.0 (atol=1e-05)
```

**Comment by mattip** [comment]
LGTM. Thanks for reformatting it into 3 lines, that is better.

**Comment by mruberry** [comment]
I also like your new error message proposal.

**Comment by mruberry** [comment]
> I also like your new error message proposal.

Brief elaboration on this topic:

It'd be really nice in the pointwise numerics tests if we could relate the indices back to the inputs. Maybe we should consider a call that returns the data needed to construct the message so it can be programmatically queried, too.

**Comment by pmeier** [comment]
@mruberry 

Note that the actual implementation of this is in the next PR of the stack #55145. I'm going to answer there.

**Comment by mruberry** [comment]
Unfortunately this part of the stack is hitting some issues being landed. Would you create single PRs or a new stack, @pmeier?

**Comment by pmeier** [comment]
Sure, can do that. Could you explain two things so I can avoid such a situation in the future?

1. Why is this PR causing problems? It just adds tests which seems to be passing fine on the public CI. Is any tests failing internally?
2. How would starting a new stack / opening a single PR help? Given that all prior PRs landed, shouldn't this PR act the same way as a new stack starting here?

**Comment by mruberry** [comment]
> Sure, can do that. Could you explain two things so I can avoid such a situation in the future?
> 
> 1. Why is this PR causing problems? It just adds tests which seems to be passing fine on the public CI. Is any tests failing internally?

I'm not totally sure. Possibly because a PR in the stack it's based on had to be reimported.

All the tests are passing, ghimport is just refusing to reconcile the PR and the diff.

> 2. How would starting a new stack / opening a single PR help? Given that all prior PRs landed, shouldn't this PR act the same way as a new stack starting here?

Because this PR seems to be "poisoned" by its stack.

File: test/test_testing.py [hunk 2]
```diff
@@ -683,5 +685,224 @@ def test_calculate_2_shards_against_optimal_shards(self):
                 self.assertEqual(sorted_tests, sorted_shard_tests)
 
 
+class TestAsserts(TestCase):
```
Comment (walterddr) [inline test/test_testing.py:688]: nit why don't we create `TestAsserts` and `TestAssertsMultiDevice` ? seems like
1. all but 2 is decorated with `@onlycuda` and 
2. seems like in the future we don't want to just check against cuda, but other device types too?

**Review by walterddr — APPROVED** [review]
looks good.

File: test/test_testing.py (hunk 2 above)
Comment (pmeier) [reply test/test_testing.py:688]: That is a good idea. Since it is not blocking I'll put it at the end of the stack to avoid fixing all the conflicts.

**Comment by mruberry** [comment]
Unfortunately our internal tools are unable to merge this, @pmeier. Would you try rebasing this PR?
